include/mfmg/dealii_operator_device.templates.cuh

```diff
@@ -22,6 +22,127 @@
 
 namespace mfmg
 {
+template <typename VectorType>
+SparseMatrixDeviceOperator<VectorType>::SparseMatrixDeviceOperator(
+    std::shared_ptr<SparseMatrixDevice<typename VectorType::value_type>> matrix)
+    : _matrix(matrix)
+{
+  ASSERT(matrix != nullptr, "The matrix must exist");
+}
+
+template <typename VectorType>
+void SparseMatrixDeviceOperator<VectorType>::apply(VectorType const &x,
+                                                   VectorType &y) const
+{
+  _matrix->vmult(y, x);
+}
+
+template <typename VectorType>
+std::shared_ptr<MatrixOperator<VectorType>>
+SparseMatrixDeviceOperator<VectorType>::transpose() const
+{
+  // TODO: to do this on the GPU
+  // Copy the data to the cpu and then, use the trilinos to compute the
+  // transpose. This is not the most efficient way to do this but it is the
+  // easiest.
+  unsigned int const local_nnz = _matrix->local_nnz();
+  unsigned int const n_local_rows = _matrix->n_local_rows();
+  std::vector<value_type> values(local_nnz);
+  std::vector<int> column_index(local_nnz);
+  std::vector<int> row_ptr(n_local_rows + 1);
+
+  // Copy the data to the host
+  cuda_mem_copy_to_host(_matrix->val_dev, values);
+  cuda_mem_copy_to_host(_matrix->column_index_dev, column_index);
+  cuda_mem_copy_to_host(_matrix->row_ptr_dev, row_ptr);
+
+  // Create the sparse matrix on the host
+  dealii::IndexSet locally_owned_rows = _matrix->locally_owned_range_indices();
+  dealii::TrilinosWrappers::SparseMatrix sparse_matrix(
+      locally_owned_rows, _matrix->locally_owned_domain_indices(),
+      _matrix->get_mpi_communicator());
+
+  std::vector<unsigned int> rows;
+  locally_owned_rows.fill_index_vector(rows);
+  for (unsigned int i = 0; i < n_local_rows; ++i)
+  {
+    unsigned int const row = rows[i];
+    unsigned int const n_cols = row_ptr[i + 1] - row_ptr[i];
+    for (unsigned int j = row_ptr[i]; j < row_ptr[i + 1]; ++j)
+      sparse_matrix.set(row, column_index[j], values[j]);
```

Comment:
Is there no way in dealii to construct a matrix from 3 pointers? Or at least row by row instead of entry by entry? This seems tedious.